### ISSUE TITLE
Jacobian info in FNO

### ISSUE THREAD

**Issue by RUSSTARUS** [Ceyron/numeric-notes#23]
Hi Felix! 

Thank you for a very good video for FNO in JAX
https://www.youtube.com/watch?v=74uwQsBTIVo

I want to ask you: how can I modify the loss function here by adding the square of the norm of the difference in the Jacobians of the network function
and our data?
I heard that adding   +lambda * || Jacob_Fnn(x) - Jacob_Fdata(x) ||^2       to the loss function improves the generalization of the neural network.

I have a rough understanding of how to write a Jacobian for a neural network function - Fnn(x)

But I don't know how to do this with data (without a data function - Fdata(x)). How to build a Jacobian from our known data?
Is this possible at all?
You could try to implement this and show the result in a new video for FNO ?
It would be very interesting to know how this can be done here in general and on JAX with this FNO operator.

Ruslan

**Comment by Ceyron** [comment]
Hi Ruslan,

Thanks for your interest in the content. You can easily take network Jacobians in JAX via

```python
jacobian = jax.jacfwd(model)(x)
```

In order to add it to the loss function in a supervised way, you would need the Jacobian of the reference data. In the specific dataset used for the video, we do not have access to the underlying numerical solver. You could try to approximate it via finite differences.
It would be easier if you had access to the simulator and could generate Jacobians for each time step. You could look into [APEBench](https://youtu.be/q8fjQ4ZFynw) and [its data-generating simulator Exponax](https://github.com/Ceyron/exponax) that runs in JAX and supports automatic differentiation.
Maybe this video on estimating the Lyapunov exponent with autodiff can be helpful: https://youtu.be/zRMBIkpcuu0

**Comment by RUSSTARUS** [comment]
Thanks Felix!

Yes, I've thought a little about all this. We have data at different grid steps.

 Is it correct to use the difference of values F(x_i)-F(x_i-1) for point (x_i)? That is, (Y_i - Y_i-1) data for the value of the function?

In this case, we can only obtain a diagonal matrix for the data.
However, the full Jacobian of the neural network function will be a full matrix.

I think it's correct that the correct neural network Jacobian matrix should also tend to a diagonal matrix.

Ruslan

**Comment by Ceyron** [comment]
> Thanks Felix!

You're welcome :)

> Is it correct to use the difference of values F(x_i)-F(x_i-1) for point (x_i)? That is, (Y_i - Y_i-1) data for the value of the function?

You can use different kinds of finite difference schemes. The backward difference you suggest is a good first start. In order to get the $J_{ij}$ entry of the Jacobian, you would have to modify a $\Delta x$ into the $j$-th direction of the input, i.e., for your backward difference

$$
J_{ij} \approx \frac{1}{\Delta x} \left( F_i(\vec{x}) - F_i(\vec{x} - \Delta x \vec{e}_j) \right)
$$

Maybe you find the [`numpy.gradient(...)`](https://numpy.org/doc/2.3/reference/generated/numpy.gradient.html) helpful.

> In this case, we can only obtain a diagonal matrix for the data. However, the full Jacobian of the neural network function will be a full matrix.

With the approach above, you will get all the entries. If your input has $N$ degrees of freedom, the Jacobian matrix will have $N \times N$ entries.

> I think it's correct that the correct neural network Jacobian matrix should also tend to a diagonal matrix.

In most cases, no. You will see sparsity in the Jacobian if the network has a limited receptive field (e.g., stacked convolutions). If you have global connectivity, e.g., via an MLP, then the Jacobian is full and dense.

**Comment by RUSSTARUS** [comment]
Hello Felix!

Interestingly, the Jacobian of the data should reflect the Markovian dependence of the data N steps back if we have N data. Right?

The Jacobian matrix of the data will be diagonal only if the data is completely independent.

I'm not very good at JAX (I might get something wrong)
I would like to see your implementation of the correction for Jacobians of the neural network in the next video for FNO. Can you do it?

I really like FNO, and when I saw your implementation of FNO in JAX, I really liked it.
I tried your FNO_JAX on my specific data and noticed an increase in speed compared to the usual implementation.
I also noticed that the optimizer (optimizer = optax.rprop(...)) for my data is an order of magnitude faster than others (I went through everything)

I also have other questions about the implementation of FNO (under JAX), but that's later...
Ruslan
